HTS221: Check for drdy_gpio defs into DTS

avisconti opened the pull request:
Added the gpio for HTS221 drdy in Device Tree.
Fix #28443

Built only, not tested.
Issue here is because HTS221 has been added in build_all trigger file, but the gpio drdy was not provided in disco_l475_iot1 device tree.
I have not tested it on the board itself, I can do tomorrow if required. Please review also the schematic, I
may have an older one. Thanks!

EDIT:
Totally changed the fix. Instead of defining the drdy_gpios in DTS (it is a not required option) I check it in the hts221 driver itself before using the generated macro.

erwango commented on boards/arm/disco_l475_iot1/disco_l475_iot1.dts:
```diff
@@ -81,6 +81,7 @@
 	hts221@5f {
 		compatible = "st,hts221";
 		reg = <0x5f>;
+		drdy-gpios = <&gpiod 15 GPIO_ACTIVE_HIGH>;
```
@avisconti drdy-gpio is described as not required in binding (cf `dts/bindings/sensor/st,hts221.yaml`)
```
properties:
    drdy-gpios:
      type: phandle-array
      required: false
      description: DRDY pin
```
If not required, it should not produce error if not present.
So, maybe this change in the board is correct, but one more fix is required:
-either this property should be made `required: true`
-either hts221 driver should not produce build error when not present

avisconti replied:
@erwango 
I think it would be better thesecond option (i.e. change the driver). But (and I had a similar question in a recent PR) I don't know if it is possible to check in the driver if any gpio is defined or not. Currently I just check CONFIG_HTS221_TRIGGER.

I explain better. We have macros like "DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)" that can be used in the driver to see if in the DT we have defined instances on SPI bus. And this is helpful for enable/disable part of code in the driver.
In case of GPIO what would be the best way to check? Maybe if DT_INST_GPIO_LABEL(inst, drdy_gpios) is NULL? But if there are multiple instances we would require to have a macro with _ANY_INST_. Do you see my point?

If we want to fix it quickly we can also remove HTS221 from the build_all trigger file. But I guess that sooner or later we need to have some special macro as described above (if not yet present)

erwango replied:
@avisconti you can use the following DT macro to hcekc presence of  an optional dt prop: `DT_INST_NODE_HAS_PROP`.
You can check examples of usage in the tree.

avisconti replied:
@erwango 
OK, I'll do in this way as it seems more correct. I will leave the DTS modification for disco_l475_iot1 for a later change, as it is a totally different context (by the way I do not think it is strictly required drdy for a humidity sensor, so we may want to carefullt evaluate it).

By the way I guess that if we require to make hts221 multi-instance we need a more complex macro with ANY in it (Does ANY hts221 instance has this PROP?). Do you agree?

ABOSTM reviewed (approved):
LGTM.
I checked Schematics and PD15 is the right pin: HTS221_DRDY_EXTI15.

erwango reviewed (changes requested):
Question on the property requirements.

Besides, it seems that `drdy-gpios` is not present in following boards/shields and will produce the same result:
-boards/arm/b_l4s5i_iot01a/b_l4s5i_iot01a.dts
-boards/shields/x_nucleo_iks01a1/x_nucleo_iks01a1.overlay
-boards/shields/x_nucleo_iks01a2/x_nucleo_iks01a2.overlay
-boards/shields/x_nucleo_iks01a3/x_nucleo_iks01a3.overlay

It would be nice to fix these all together.

avisconti commented:
@erwango @MaureenHelm 
I have to leave now, and tomorrow I can spend some more time.
I pushed a new version but I'm not satisfied. I'm mixing both CONFIG_HTS221_TRIGGER and checks for DT_INST_NODE_HAS_PROP(0, drdy_gpios). Not sure I can do any better.

pabigot reviewed (changes requested):
If the devicetree provides `drdy-gpios` but the driver isn't configured to use it this approach wastes both data and code space.

I think a better solution would be to continue to use `CONFIG_HTS221_TRIGGER` as the primary flag to control references to DRDY gpio use, but to add:

    BUILD_ASSERT(DT_INST_NODE_HAS_PROP(0, drdy_gpios) || !IS_ENABLED(CONFIG_HTS221_TRIGGER),
                 "Cannot enable trigger without drdy-gpios");

It would also be nice to filter the sample based on that, but I don't think we have the capability (need `dt_prop_exists` in the filter language).

Alternatively:

    #define HTS221_TRIGGER_ENABLED (DT_INST_NODE_HAS_PROP(0, drdy_gpios) && IS_ENABLED(CONFIG_HTS221_TRIGGER))

in the header and use `HTS221_TRIGGER_ENABLED` as the standard condition.

avisconti commented:
@pabigot 
I think they should remain separated because, in case of multi-instance (sooner or later I will introduce it in HTS221), we need to check the presence of drdy_gpios instance by instance. How can we do that?

Probably we need some filter that currently is not existing.

What about

    #define HTS221_TRIGGER_ENABLED(inst) (DT_INST_NODE_HAS_PROP(inst, drdy_gpios) && IS_ENABLED(CONFIG_HTS221_TRIGGER))

avisconti commented:
@pabigot 
See for example the current lsm6dsl.c, it is a multi-instance and we have exactly the same issue, i.e. if irq_gpios is not defined in DT then the build fails. How can we change that code? I was thinking to do a change similar to what I'm proposing here, but again there should be a better way.

pabigot commented:
> I think they should remain separated because, in case of multi-instance (sooner or later I will introduce it in HTS221), we need to check the presence of drdy_gpios instance by instance. How can we do that?

Yeah, OK.  I was proposing a path that would get #28443 fixed for 2.4.0.

For multi instance the preprocessor check is solely dependent on `CONFIG_HTS221_TRIGGER`, and the runtime check looks for a non-null DRDY device name.  Then the instantiation macros use `COND_CODE_1()` to check for the devicetree property to conditionally initialize the configuration structure fields related to the GPIO.  Something like this:
```
``
#ifdef CONFIG_LSM6DSL_TRIGGER
#define LSM6DSL_CFG_IRQ(inst) \
                .irq_dev_name = DT_INST_GPIO_LABEL(inst, irq_gpios),    \
                .irq_pin = DT_INST_GPIO_PIN(inst, irq_gpios),           \
                .irq_flags = DT_INST_GPIO_FLAGS(inst, irq_gpios),
#else
#define LSM6DSL_CFG_IRQ()
#endif

#define LSM6DSL_CONFIG_I2C(inst)                                        \
        {                                                               \
                .bus_name = DT_INST_BUS_LABEL(inst),                    \
                .bus_init = lsm6dsl_i2c_init,                           \
                .bus_cfg = { .i2c_slv_addr = DT_INST_REG_ADDR(inst), }, \
		COND_CODE_1(DT_INST_HAS_PROP(inst, irq_gpios),          \
		            (CONFIG_LSMDSL_TRIGGER(inst)),())           \
        }
```

Note this gets rid of the duplication of the non-conditional initialization which cleans the code up nicely (you don't need two versions of `LSM6DSL_CONFIG_I2C` based on `CONFIG_LSM6DSL_TRIGGER`).

Does that make sense?

avisconti commented:
> > I think they should remain separated because, in case of multi-instance (sooner or later I will introduce it in HTS221), we need to check the presence of drdy_gpios instance by instance. How can we do that?
> 
> Yeah, OK. I was proposing a path that would get #28443 fixed for 2.4.0.
> 
> For multi instance the preprocessor check is solely dependent on `CONFIG_HTS221_TRIGGER`, and the runtime check looks for a non-null DRDY device name. Then the instantiation macros use `COND_CODE_1()` to check for the devicetree property to conditionally initialize the configuration structure fields related to the GPIO. Something like this:
> 
> ```
> ``
> #ifdef CONFIG_LSM6DSL_TRIGGER
> #define LSM6DSL_CFG_IRQ(inst) \
>                 .irq_dev_name = DT_INST_GPIO_LABEL(inst, irq_gpios),    \
>                 .irq_pin = DT_INST_GPIO_PIN(inst, irq_gpios),           \
>                 .irq_flags = DT_INST_GPIO_FLAGS(inst, irq_gpios),
> #else
> #define LSM6DSL_CFG_IRQ(inst)
> #endif
> 
> #define LSM6DSL_CONFIG_I2C(inst)                                        \
>         {                                                               \
>                 .bus_name = DT_INST_BUS_LABEL(inst),                    \
>                 .bus_init = lsm6dsl_i2c_init,                           \
>                 .bus_cfg = { .i2c_slv_addr = DT_INST_REG_ADDR(inst), }, \
> 		COND_CODE_1(DT_INST_HAS_PROP(inst, irq_gpios),          \
> 		            (CONFIG_LSMDSL_TRIGGER(inst)),())           \
>         }
> ```
> 
> Note this gets rid of the duplication of the non-conditional initialization which cleans the code up nicely (you don't need two versions of `LSM6DSL_CONFIG_I2C` based on `CONFIG_LSM6DSL_TRIGGER`).
> 
> Does that make sense?

Yes it does.
I need time to make changes here, but definetely makes sense.
Thanks

avisconti committed:
drivers: sensor: hts221: Add checks for drdy_gpios macros

To avoid build errors enable the trigger only if DRDY gpio is
defined in DTS. Fix #28443

Signed-off-by: Armando Visconti <armando.visconti@st.com>

erwango commented on drivers/sensor/hts221/hts221.h:
```diff
@@ -12,6 +12,9 @@
 #include <zephyr/types.h>
 #include <drivers/gpio.h>
 
+#define HTS221_TRIGGER_ENABLED (DT_INST_NODE_HAS_PROP(0, drdy_gpios) && \
```
Shouldn't we add a check to be sure that `drdy_gpios` is present when `CONFIG_HTS221_TRIGGER` is enabled?

pabigot replied:
I don't think so.  The Kconfig signals that support for the trigger should be present in the driver.  That support doesn't require that the signal actually be present in the hardware.  (It might not be in a shield, and the application should get a runtime error if it the code tries to use something that isn't supported by the hardware.)

We probably don't have any policy on that, so I suppose we could start to require that if the driver enables interrupts the hardware must provide support for them.  I don't think that's a good idea, though.

erwango replied:
ok, txs for feedback

erwango reviewed (dismissed):
Tested ok on disco_l475_iot1

pabigot reviewed (changes requested):
Looks good, though there's still some reference that isn't getting excluded causing the build failures.  So will approve once it passes CI.

pabigot commented:
Approved now to remove any delay for re-review after this is updated to pass CI.  I don't expect those changes to be a problem.

avisconti commented:
> Looks good, though there's still some reference that isn't getting excluded causing the build failures. So will approve once it passes CI.

Yeah, thanks.
A couple of silly mistakes in trigger file. But nothing too serious. Should pass now.
